gpflow/base.py

```diff
@@ -1,51 +1,41 @@
 import functools
 from enum import Enum
-from typing import List, Optional, Tuple, Union
+from typing import Any, List, Optional, Sequence, Tuple, TypeVar, Union
 
 import numpy as np
 import tensorflow as tf
 import tensorflow_probability as tfp
+from IPython.lib import pretty
```

Comment:
Can we please not add IPython as a strict dependency?

Reply:
does this mean `_repr_pretty_` is an optional feature? if so, we should mark it as such

Reply:
that is, whatever calls `_repr_pretty_` must have IPython as a dependency to be able to construct the `p` argument, so IPython is a dependency even if we don't import it explicitly. So I've taken 'strict' to mean that not all users will use `_repr_pretty_` and shouldn't require the dependency if they don't use it. Is that what you mean?

Reply:
if by strict you mean what I've detailed, I'd like to have a wider discussion on what the GPflow standard is wrt optional dependencies. I'm wary of adding lots as it sounds like a serious maintenance headache

Reply:
The only thing that calls `_repr_pretty_` is IPython itself, and it's marked private, so I can't see any issues in just leaving it as it is. If you don't have IPython, there is no reason for it being called. If you do have IPython, you transparently get much nicer object-prints. I agree that if we had lots and lots of optional dependencies, we should reconsider, but at this point I think it's fine as it is.

Reply:
one main benefit of typing it is that we gain type checking for our implementation of `_repr_pretty_`. I feel that omitting types cos something's an optional dependency would be somewhat of an antipattern. If it's an optional dependency, there are other ways of allowing for that

Reply:
I think I've solved it: conditionally import IPython only for type checking